third_party/move/move-compiler-v2/src/pipeline/instruction_reordering.rs

```diff
@@ -0,0 +1,771 @@
+    fn adjusted_ordering(
+        &self,
+        offsets: Vec<CodeOffset>,
+        node: CodeOffset,
+        ref_args: &BTreeMap<TempIndex, BTreeSet<CodeOffset>>,
+        reads: &BTreeMap<TempIndex, BTreeSet<CodeOffset>>,
+        prepares: &BTreeMap<CodeOffset, TempIndex>,
+    ) -> Option<(Vec<CodeOffset>, BTreeSet<CodeOffset>)> {
+        assert!(offsets.contains(&node));
+        // Check that for each ordered pair of DFS-ordered offsets starting from `node`,
+        // all the constraints are satisfied.
+        for i in 0..offsets.len() {
+            for j in i + 1..offsets.len() {
+                let before = offsets[i];
+                let after = offsets[j];
+                if self
+                    .constraints
+                    .get(&after)
+                    .is_some_and(|nodes| nodes.contains(&before))
+                {
+                    return None;
+                }
+            }
+        }
+        // Check that for each DFS-unvisited offset that lies between any of the visited
+        // offsets and `node`, all the constraints are satisfied when unvisited offsets
+        // are moved above all visited offsets.
+        // These checks do not have to consider `Prepare` instructions.
+        let visited = offsets.iter().collect::<BTreeSet<_>>();
+        let min_visited = **visited
+            .first()
+            .expect("at least one offset must be visited");
+        for before in min_visited..=node {
+            if !visited.contains(&before) {
+                // Is it safe to move `before` before everything else?
+                for after in offsets.iter() {
+                    if self
+                        .constraints
+                        .get(after)
+                        .map_or(false, |nodes| nodes.contains(&before))
+                    {
+                        return None;
+                    }
+                }
+            }
+        }
+        // Additional checks and restrictions corresponding to the placement of
+        // `Prepare` instructions.
+        let mut skip_prepares = BTreeSet::new();
+        let mut copy_only_prepares = BTreeSet::new();
+        for (i, offset) in offsets.iter().enumerate() {
+            if let Some(prepared_tmp) = prepares.get(offset) {
+                // For each `Prepare tmp` instruction, check if:
+                // 1. Until its use, whether there are any reference-related instructions
+                //    targeting `tmp`. In such a case, we have to skip this `Prepare`
+                //    instruction, because we cannot eagerly access `tmp`.
+                // 2. Until its use, whether there are any reads of `tmp`. In such a case,
+                //    we have to notify that this `Prepare` instruction can only result in
+                //    a copy of `tmp` (or a no-op, which is always safe for `Prepare`
+                //    instructions). If it results in a move, then a subsequent read of
```

Comment:
What about a case like
```
    t2 = ...( ... t1 ...);
    operation(t1, t2);
```
where the operation is the last use of `t1`?  This rule 2 says we wind up doing:
```
    Prepare(copy, t1);
    t2 = ... t1 ...;
    operation(t1, t2);
```
Will the CG be able to turn the use of `t1` to compute `t2` into a `move`?